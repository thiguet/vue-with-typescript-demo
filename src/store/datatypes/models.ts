interface Unique {
    id?: string;
}

export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product extends Unique {
    name: string;
    measure: Measures;
    qtd: number;
    minQtd: number;
    image?: string;
}

export interface User extends Unique {
    name: string;
    email: string;
}

export interface ReportItem {
    text: string;
    value: number;
}

export type ReportData = ReportItem[];

export interface QuantityReportData {
    productName: string;
    productQuantity: number;
}

export type QuantityReport = QuantityReportData[];
export interface MeasuresReportData {
    measureName: string;
    measureQuantity: number;
}

export type MeasuresReport = MeasuresReportData[];

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
    reports = 'reports',
}
