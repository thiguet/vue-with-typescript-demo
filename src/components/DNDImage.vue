<template>
    <div
        class="drop img-container"
        :class="getClasses"
        @dragover.prevent="dragOver"
        @dragleave.prevent="dragLeave"
        @drop.prevent="drop($event)"
    >
        <img
            id="img"
            :class="{ fallback: !imageSource }"
            :src="imageSource ? imageSource : fallbackImage"
        />
        <h1 id="wrong-file-header" v-if="wrongFile">Arquivo inv√°lido!</h1>
    </div>
</template>

<script lang="ts">
import { DNDImageComponent } from '@/views/models.d';
import Vue from 'vue';
import Component from 'vue-class-component';

@Component({
    name: 'DNDImage',
})
export default class DNDImage extends Vue implements DNDImageComponent {
    isDragging: boolean;

    wrongFile: boolean;

    imageSource?: ArrayBuffer | string | null;

    fallbackImage: string;

    constructor() {
        super();
        this.isDragging = false;
        this.wrongFile = false;
        this.imageSource = null;
        this.fallbackImage = '/assets/icons/upload-image.svg';
    }

    get getClasses() {
        return { isDragging: this.isDragging };
    }

    dragOver() {
        this.isDragging = true;
    }

    dragLeave() {
        this.isDragging = false;
    }

    drop(event: { dataTransfer: DataTransfer }) {
        const { files } = event.dataTransfer;

        this.wrongFile = false;

        if (files.length === 1) {
            const file = files[0];

            if (file.type.indexOf('image/') >= 0) {
                const reader = new FileReader();
                reader.onload = (ev: ProgressEvent<FileReader>) => {
                    const fileReaderObj = ev.target as FileReader;
                    this.imageSource = fileReaderObj.result;
                    this.isDragging = false;
                };
                reader.readAsDataURL(file);
            } else {
                this.wrongFile = true;
                this.imageSource = null;
                this.isDragging = false;
            }
        }
    }
}
</script>

<style scoped>
h1 {
    font-size: 14px;
}

.drop {
    width: 100%;
    height: 100%;
    background-color: #fff;
    border: 10px solid #fff;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 4px 2px;
    transition: background-color 0.2s ease-in-out;
    font-family: sans-serif;
}

.isDragging {
    background-color: #999;
    border-color: #fff;
}
img {
    max-height: 200px;
    border-radius: 50%;
}

img.fallback {
    width: 70%;
    border-radius: 0;
}

.img-container {
    display: flex;
    position: relative;
    width: 200px;
    height: 200px;
    padding: 10px;
    border-radius: 50%;
}

@media (max-width: 320px) {
    .img-container {
        width: 150px;
        height: 150px;
    }
}
</style>
