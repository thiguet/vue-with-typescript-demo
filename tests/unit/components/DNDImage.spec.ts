import { mount } from '@vue/test-utils';
import DNDImage from '@/components/DNDImage.vue';
import faker from 'faker';
import { DNDImageComponent, DNDImageData } from '@/views/models.d';
import { DNDImageProps, ImageMimeTypes } from './models';

describe('DNDImage.vue', () => {
    let data: DNDImageData;
    let props: DNDImageProps;

    const getFakeFile = (type: string) => {
        const name = faker.system.fileName();
        const file = new File(['(⌐□_□)'], name, {
            type,
        });

        return {
            name,
            type,
            file,
        };
    };

    const build = () => {
        const wrapper = mount(DNDImage, {
            propsData: { ...props },
        });

        const DNDImageComp = () => {
            const DNDImageInstance = (wrapper.vm as unknown) as DNDImageComponent;

            jest.spyOn(DNDImageInstance, 'dragLeave');
            jest.spyOn(DNDImageInstance, 'dragOver');
            jest.spyOn(DNDImageInstance, 'drop');

            return DNDImageInstance;
        };

        return {
            wrapper,
            DNDImageComp,
            img: () => wrapper.find('#img'),
            imgContainer: () => wrapper.find('.img-container'),
            wrongFileHeader: () => wrapper.find('#wrong-file-header'),
        };
    };

    beforeEach(() => {
        props = {
            id: faker.random.uuid(),
            value: faker.image.image(),
            setValue: jest.fn(),
        };
    });

    it('renders component', async () => {
        props.id = 'someFixedId';
        props.value = 'someFixedValue';

        const { wrapper } = build();

        expect(wrapper).toMatchSnapshot();
    });

    it('renders main components', () => {
        const { img, imgContainer, wrongFileHeader } = build();

        expect(imgContainer().exists()).toBe(true);
        expect(img().exists()).toBe(true);
        expect(wrongFileHeader().exists()).toBe(false);
    });

    it('display wrong answer based on component data.', async () => {
        const { wrapper, img, imgContainer, wrongFileHeader } = build();

        data = {
            wrongFile: true,
            isDragging: false,
        };

        await wrapper.setData({ ...data });

        expect(imgContainer().exists()).toBe(true);
        expect(img().exists()).toBe(true);
        expect(wrongFileHeader().exists()).toBe(true);
    });

    it('display image based on component data.', async () => {
        const { wrapper, img, imgContainer, wrongFileHeader } = build();

        data = {
            imageSource: 'http://lorempixel.com/640/480/food',
            wrongFile: false,
            isDragging: false,
        };

        await wrapper.setData({ ...data });

        expect(imgContainer().exists()).toBe(true);
        expect(img().exists()).toBe(true);
        expect(wrongFileHeader().exists()).toBe(false);
    });

    it('simulate a dragOver event.', async () => {
        const { DNDImageComp, imgContainer } = build();

        const options = { preventDefault: jest.fn() };
        const event = new Event('dragover');
        Object.assign(event, options);
        await imgContainer().element.dispatchEvent(event);

        await setTimeout(() => {
            expect(options.preventDefault).toBeCalled();
            expect(DNDImageComp().isDragging).toBe(true);
        }, 0);
    });

    it('simulate a dragLeave event.', async done => {
        const { DNDImageComp, imgContainer } = build();

        const event = new Event('dragleave');

        const options = {
            preventDefault: jest.fn(),
        };

        Object.assign(event, options);

        await imgContainer().element.dispatchEvent(event);

        await setTimeout(() => {
            expect(options.preventDefault).toBeCalled();
            expect(DNDImageComp().isDragging).toBe(false);
            done();
        }, 0);
    });

    it('simulate a drop event with an image.', async done => {
        const imageMimeType = faker.random.arrayElement(
            Object.values(ImageMimeTypes),
        );
        const { file } = getFakeFile(imageMimeType);
        const { DNDImageComp, imgContainer, wrapper } = build();
        interface DataTranferMock {
            files: File[];
        }
        const dataTransfer: DataTranferMock = {
            files: [file],
        };
        const options = {
            preventDefault: jest.fn(),
            dataTransfer,
        };

        const event = new Event('drop');

        Object.assign(event, options);

        imgContainer().element.dispatchEvent(event);

        // Haven't found a way of waiting DOM to trigger my event.
        // Even using setTimeout or wrapper.vm.nextTick several times, hasn't played out.
        await setTimeout(() => {
            expect(options.preventDefault).toBeCalled();
            expect(DNDImageComp().imageSource).toBeTruthy();
            expect(wrapper.emitted('change')).toBeTruthy();
            done();
        }, 1000);
    });

    it('simulate a drop event without a file.', async done => {
        const { DNDImageComp, imgContainer } = build();

        const options = {
            preventDefault: jest.fn(),
            dataTransfer: { files: [] },
        };

        const { imageSource, wrongFile, isDragging } = DNDImageComp();
        const event = new Event('drop');

        Object.assign(event, options);

        imgContainer().element.dispatchEvent(event);

        // Haven't found a way of waiting DOM to trigger my event.
        // Even using setTimeout or wrapper.vm.nextTick several times, hasn't played out.
        await setTimeout(() => {
            expect(options.preventDefault).toBeCalled();
            // Expecting them to be the same as before.
            expect(DNDImageComp().imageSource).toBe(imageSource);
            expect(DNDImageComp().wrongFile).toBe(wrongFile);
            expect(DNDImageComp().isDragging).toBe(isDragging);
            done();
        }, 1000);
    });

    it('simulate a drop event with a file that is not an image.', async done => {
        const notAnImageMimeType = 'application/pdf';
        const { file } = getFakeFile(notAnImageMimeType);
        const { DNDImageComp, imgContainer } = build();
        interface DataTranferMock {
            files: File[];
        }
        const dataTransfer: DataTranferMock = {
            files: [file],
        };
        const options = {
            preventDefault: jest.fn(),
            dataTransfer,
        };

        const event = new Event('drop');

        Object.assign(event, options);

        imgContainer().element.dispatchEvent(event);

        // Haven't found a way of waiting DOM to trigger my event.
        // Even using setTimeout or wrapper.vm.nextTick several times, hasn't played out.
        await setTimeout(() => {
            expect(options.preventDefault).toBeCalled();
            expect(DNDImageComp().imageSource).toBe(null);
            expect(DNDImageComp().wrongFile).toBe(true);
            expect(DNDImageComp().isDragging).toBe(false);
            done();
        }, 1000);
    });
});
