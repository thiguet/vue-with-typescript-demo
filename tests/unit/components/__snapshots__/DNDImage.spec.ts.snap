// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DNDImage.vue renders component 1`] = `
<div class="drop img-container" id="someFixedId" value="someFixedValue" setvalue="function mockConstructor() {
        return fn.apply(this, arguments);
      }"><img id="img" src="./assets/icons/upload-image.svg" class="fallback">
  <!---->
</div>
`;
